api/src/metrics/lib/create-metrics.ts

```diff
@@ -0,0 +1,303 @@
+import { useEnv } from '@directus/env';
+import { toArray } from '@directus/utils';
+import { randomUUID } from 'node:crypto';
+import { Readable } from 'node:stream';
+import { promisify } from 'node:util';
+import pm2 from 'pm2';
+import type { MetricObjectWithValues, MetricValue } from 'prom-client';
+import { AggregatorRegistry, Counter, Histogram, register } from 'prom-client';
+import { getCache } from '../../cache.js';
+import { hasDatabaseConnection } from '../../database/index.js';
+import { redisConfigAvailable, useRedis } from '../../redis/index.js';
+import { getStorage } from '../../storage/index.js';
+
+const isPM2 = 'PM2_HOME' in process.env;
+const METRICS_SYNC_PACKET = 'directus:metrics---data-sync';
+
+const listApps = promisify(pm2.list.bind(pm2));
+const sendDataToProcessId = promisify(pm2.sendDataToProcessId.bind(pm2));
+
+export function createMetrics() {
+	const env = useEnv();
+
+	const aggregates = new Map();
+
+	/**
+	 * Listen for PM2 metric data sync messages and add them to the aggregate
+	 */
+	if (isPM2) {
+		process.on('message', (packet: any) => {
+			if (!packet.data || packet.topic !== METRICS_SYNC_PACKET) return;
+
+			aggregate(packet.data);
+		});
+	}
+
+	async function generate() {
+		const checkId = randomUUID();
+
+		await Promise.all([
+			trackDatabaseMetric(),
+			trackCacheMetric(checkId),
+			trackRedisMetric(checkId),
+			trackStorageMetric(checkId),
+		]);
+
+		/**
+		 * Sync generated metrics with all pm2 instances
+		 */
+		if (isPM2) {
+			try {
+				// sync
+				const apps = await listApps();
+
+				const data = await register.getMetricsAsJSON();
+
+				const syncs = [];
+
+				for (const app of apps) {
+					if (app.pm_id === undefined || app.pid === 0 || app.name !== 'directus') {
+						continue;
+					}
+
+					syncs.push(
+						sendDataToProcessId(app.pm_id, {
+							data: { pid: process.pid, metrics: data },
+							topic: METRICS_SYNC_PACKET,
+						}),
+					);
+				}
+
+				await Promise.allSettled(syncs);
+			} catch {
+				// ignore
+			}
+		}
+	}
+
+	/**
+	 * Add PM2 synced metric to the aggregate store.
+	 * Subsequent syncs for the given instance will override previous value.
+	 */
+	async function aggregate(data: { pid: number; metrics: MetricObjectWithValues<MetricValue<string>>[] }) {
+		aggregates.set(data.pid, data.metrics);
+	}
+
+	async function readAll(): Promise<string> {
+		/**
+		 * In a PM2 context we must aggregate the metrics across instances ensuring
+		 * only currently active instances are added to the aggregate
+		 */
+		if (isPM2 && aggregates.size !== 0) {
+			const apps = await listApps();
+
+			const aggregate = [];
+
+			for (const app of apps) {
+				if (aggregates.has(app.pid)) {
+					aggregate.push(aggregates.get(app.pid));
+				}
+			}
+
+			if (aggregate.length !== 0) {
+				return AggregatorRegistry.aggregate(aggregate).metrics();
```

Comment:
Nice way to reuse the `AggregatorRegistry`'s functionality without using their transport method.

Reply:
Thanks :)